File: JavaScript/763-Partition Labels.js
/**
 * @param {string} S
 * @return {number[]}
 */
var partitionLabels = function(S) {
    let hmap = {};
    
    for (let i=0; i<S.length; i++){
        hmap[S[i]] = i;
    }
    
    let start = 0, end = 0;
    let res = [];
    for (let j=0; j<S.length; j++){
        end = Math.max(end, hmap[S[j]]);
        if (j === end){
            res.push(end-start+1);
            start = end + 1;
        }
    }
    return res;
};

File: C++/763-Partition Labels.cpp
class Solution {
public:
    vector<int> partitionLabels(string s) {
        vector<int> last(26, -1);
        int N = s.size();

        for (int i=0; i<N; i++) {
            last[s[i]-'a'] = i;
        }

        vector<int> res;
        int left = 0, maxEnd = 0;
        for (int i=0; i<N; i++) {
            maxEnd = max(maxEnd, last[s[i]-'a']);
            if (i == maxEnd) {
                res.push_back(i-left+1);
                left = i+1;
            }
        }

        return res;
    }
};